Assets/Scripts/Augment/BulletModifiers/DecalModifier.cs

```diff
@@ -0,0 +1,65 @@
+using UnityEditor.Experimental.GraphView;
+using UnityEngine;
+using UnityEngine.Rendering.Universal;
+
+enum MarkMode 
+{
+    ON_HIT,
+    ON_RICOCHET
+}
+
+public class DecalModifier : MonoBehaviour, ProjectileModifier
+{
+    [SerializeField] private MarkMode mode = MarkMode.ON_HIT;
+
+    // Note that the decal has to be quite deep (like 1 unit or so) because of the somewhat imprecise colliders we have
+    [SerializeField] private DecalProjector decal;
+
+    [Range(.2f, .8f)] [SerializeField] private float depthOffsetFraction = .6f;
+
+    [Range(0f, 180f)] [SerializeField] private float angleVariation = 180f;
+
+    [SerializeField] private float lifetime = 60;
+
+    public void Attach(ProjectileController projectile)
+    {
+        if (mode == MarkMode.ON_HIT)
+        {
+            projectile.OnColliderHit += OnHit;
+        }
+        else
+        {
+            projectile.OnRicochet += OnHit;
+        };
+    }
+
+    public void Detach(ProjectileController projectile)
+    {
+        if (mode == MarkMode.ON_HIT)
+        {
+            projectile.OnColliderHit -= OnHit;
+        }
+        else
+        {
+            projectile.OnRicochet -= OnHit;
+        };
+    }
+
+    private void OnHit(Collider target, ref ProjectileState state)
+    {
+        // Place the decal some distance away from the target so that it is more likely to be projected onto a surface
+        var depthOffset = decal.size.z * depthOffsetFraction;
+        var position = target.ClosestPoint(state.position) - state.direction * depthOffset;
+
+        var spawnedDecal = Instantiate(decal, position, state.rotation);
+        spawnedDecal.transform.parent = target.transform;
+
+        // TODO optionally force it to point down towards the target (requires surface normal)
+        // (this should fix the look of explosions at least)
```

Comment:
Maybe we should clamp the angle of which these decals can be projected at the surface from no matter the projectile type?
We might be able to avoid creating too long decal lines (like picture below) when the state.direction converges towards being parallel to the surface it is projected on by clamping. This does of course need the surface normal, but the ricochet modifier has some normal aproximation code I wrote that might just do the trick in this case.

![image](https://github.com/hackerspace-ntnu/Red-Planet-Rampage/assets/54811121/455c8fcd-a402-4813-90a6-116bb329c8d2)

Reply:
Wellllll ummmm check the updated code 😅